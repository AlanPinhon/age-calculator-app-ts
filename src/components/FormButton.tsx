export const FormButton = () => {
  return (
    <>
      <button type="submit">
        <img className="arrow-btn" src="../src/assets/images/icon-arrow.svg" alt="icon-arrow"/>
      </button>
    </>
  )
}