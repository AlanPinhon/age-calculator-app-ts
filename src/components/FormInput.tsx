import { ChangeEvent} from "react";

type inputProps = {
  labelId: string;
  placeholderText:string;
  setInputValue: (value:string) => void;
  error:string
}

export const FormInput = ({labelId, placeholderText, setInputValue, error}:inputProps) => {

  const handleInputValue = ({target}:ChangeEvent<HTMLInputElement>) => {
    const newValue = target.value;
    setInputValue(newValue);
  }

  return (
    <>
      <section className="container-input">
        <label htmlFor={labelId}>{labelId}</label>
        <input
          className={error && "input-error"}
          type="text"
          id={labelId}
          onChange={handleInputValue}
          placeholder={placeholderText}
        />
        {error && <p className="msg-error">{error}</p>}
      </section>
    </>
  )
};