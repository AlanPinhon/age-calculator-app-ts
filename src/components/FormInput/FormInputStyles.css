.container-input{
  display: flex;
  flex-direction: column;
  width: 22%;
  height: 50%;
  margin-right: 2rem;

  & label{
    font-weight: 700;
    letter-spacing: .25rem;
    text-transform: uppercase;
    font-size: .8rem;
    color: var(--smokey-grey);
    margin-bottom: .5rem;
  }

  & input{
    font-family: 'Poppins', sans-serif;
    font-weight: 700;
    border: 1px solid var(--light-grey);
    border-radius: .5rem;
    font-size: var(--font-size-inputs);
    line-height: 4rem;
    height: 4rem;
    padding-left: 1rem;
    margin-bottom: .5rem;

    &::placeholder{
      color: rgb(149, 145, 145);
    }

    &:focus{
      outline: none;
      border-color: var(--purple)
    }
  }

  
  .input-error{
    border: 1px solid var(--light-red)
  }

  .label-error{
    color: var(--light-red);
  }
}

.msg-error{
  color: var(--light-red);
  font-size: .85rem;
  font-style: italic;
}

@media screen and (max-width: 850px){

  .container-input{
    width: 30%;
    margin-right: 0;

    & input{
      font-size: 1.25rem;
      padding-left: .85rem;
      line-height: 3.5rem;
      height: 3.35rem;
    }
  }

}