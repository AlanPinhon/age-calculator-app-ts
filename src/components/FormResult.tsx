type resultsProps = {
  ageResults: { years: number; months: number; days: number } | null;
}

export const FormResult = ({ageResults}:resultsProps) => {

  const years = ageResults?.years;
  const months = ageResults?.months;
  const days = ageResults?.days;

  return (
    <>
      <section className="container-result">
        <h2 className="date-result"><span className="date-number">{years ?? '--'}</span> years</h2>
        <h2 className="date-result"><span className="date-number">{months ?? '--'}</span> months</h2>
        <h2 className="date-result"><span className="date-number">{days ?? '--'}</span> days</h2>
      </section>
    </>
  )
}