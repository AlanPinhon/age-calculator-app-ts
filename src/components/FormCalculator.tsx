import { useState } from "react";
import { FormInput } from "./FormInput";


export const FormCalculator = () => {

  const [day, setDay] = useState('');
  const [month, setMonth] = useState('');
  const [year, setYear] = useState('');

  const calculateAge = () => {
    const birthDate = new Date(Number(year), Number(month) - 1, Number(day));
    const currentDate = new Date();

    let years = currentDate.getFullYear() - birthDate.getFullYear();
    let months = currentDate.getMonth() - birthDate.getMonth();
    let days = currentDate.getDate() - birthDate.getDate();

    if (days < 0) {
      const lastMonth = new Date(
        currentDate.getFullYear(),
        currentDate.getMonth(),
        0
      );
      days += lastMonth.getDate(); // Ajustamos los días negativos
      months--;
    }

    if (months < 0) {
      years--;
      months += 12;
    }

    console.log(`${years} años, ${months} meses y ${days} días`);
  };

  return (
    <>
      <FormInput setInputValue={setDay} placeholderText="DD" inputId="Day"/>
      <FormInput setInputValue={setMonth} placeholderText="MM" inputId="Month"/>
      <FormInput setInputValue={setYear} placeholderText="YYYY" inputId="Year"/>
      <button onClick={calculateAge}>
        <img className="arrow-btn" src="../src/assets/images/icon-arrow.svg" alt="icon-arrow"/>
      </button>
    </>
  )
}