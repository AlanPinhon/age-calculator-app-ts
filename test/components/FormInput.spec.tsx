import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { FormInput } from '../../src/components/FormInput';
import { vi } from 'vitest';

describe('Tests on <FormInput />', () => {

  const placeholderTxt:string = 'DD';
  const labelId:string = 'Day';
  const error:string = 'Must be a valid day';
  const setInputValueMock = vi.fn();

  test('should to be match to the snapshot', () => {

    const {container} = render(
      <FormInput 
        placeholderText={placeholderTxt}
        labelId={labelId}
        setInputValue={setInputValueMock}
        error={error}
      />
    );

    expect(container).toMatchSnapshot();

  });

  test('should change the value in the input', () => {

    render(
      <FormInput 
        placeholderText={placeholderTxt}
        labelId={labelId}
        setInputValue={setInputValueMock}
        error={error}
      />
    );

    const inputElement = screen.getByPlaceholderText(placeholderTxt) as HTMLInputElement;
    fireEvent.change(inputElement,{target: {value: '22'}});
    expect(inputElement.value).toBe('22');

    expect(setInputValueMock).toHaveBeenCalledWith('22');
  });
  
});