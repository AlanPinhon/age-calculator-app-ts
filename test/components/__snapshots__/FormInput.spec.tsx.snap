// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests on <FormInput /> > should to be match to the snapshot 1`] = `
<div>
  <label
    for="Day"
  >
    Day
  </label>
  <input
    id="Day"
    placeholder="DD"
    type="text"
  />
  <p
    class="msg-error"
  >
    Must be a valid day
  </p>
</div>
`;
